src/libsyntax/ext/quote.rs

```diff
@@ -683,27 +723,60 @@ fn mk_tt(cx: &ExtCtxt, tt: &ast::TokenTree) -> Vec<P<ast::Stmt>> {
+            mk_tt(cx, &delimed.open_tt(), matcher).into_iter()
+                .chain(delimed.tts.iter().flat_map(|tt| mk_tt(cx, tt, matcher).into_iter()))
+                .chain(mk_tt(cx, &delimed.close_tt(), matcher).into_iter())
                 .collect()
         },
-        ast::TtSequence(..) => panic!("TtSequence in quote!"),
-    }
-}
-
-fn mk_tts(cx: &ExtCtxt, tts: &[ast::TokenTree]) -> Vec<P<ast::Stmt>> {
-    let mut ss = Vec::new();
-    for tt in tts {
-        ss.extend(mk_tt(cx, tt).into_iter());
+        ast::TtSequence(sp, ref seq) => {
+            if matcher {
+                let e_sp = cx.expr_ident(sp, id_ext("_sp"));
+
+                let stmt_let_tt = cx.stmt_let(sp, true, id_ext("tt"), cx.expr_vec_ng(sp));
+                let mut tts_stmts = vec![stmt_let_tt];
+                tts_stmts.extend(mk_tts(cx, &seq.tts[], matcher).into_iter());
+                let e_tts = cx.expr_block(cx.block(sp, tts_stmts,
+                                                       Some(cx.expr_ident(sp, id_ext("tt")))));
+                let e_separator = match seq.separator {
+                    Some(ref sep) => cx.expr_some(sp, mk_token(cx, sp, sep)),
+                    None => cx.expr_none(sp),
+                };
+                let e_op = match seq.op {
+                    ast::ZeroOrMore => mk_ast_path(cx, sp, "ZeroOrMore"),
+                    ast::OneOrMore => mk_ast_path(cx, sp, "OneOrMore"),
+                };
+                let fields = vec![cx.field_imm(sp, id_ext("tts"), e_tts),
+                                  cx.field_imm(sp, id_ext("separator"), e_separator),
+                                  cx.field_imm(sp, id_ext("op"), e_op),
+                                  cx.field_imm(sp, id_ext("num_captures"),
+                                                   cx.expr_usize(sp, seq.num_captures))];
+                let seq_path = vec![id_ext("syntax"), id_ext("ast"), id_ext("SequenceRepetition")];
+                let e_seq_struct = cx.expr_struct(sp, cx.path_global(sp, seq_path), fields);
+                let e_rc_new = cx.expr_call_global(sp, vec![id_ext("std"),
+                                                            id_ext("rc"),
+                                                            id_ext("Rc"),
+                                                            id_ext("new")],
+                                                       vec![e_seq_struct]);
+                let e_tok = cx.expr_call(sp,
+                                         mk_ast_path(cx, sp, "TtSequence"),
+                                         vec!(e_sp, e_rc_new));
+                let e_push =
+                    cx.expr_method_call(sp,
+                                        cx.expr_ident(sp, id_ext("tt")),
+                                        id_ext("push"),
+                                        vec!(e_tok));
+                vec!(cx.stmt_expr(e_push))
+            } else {
+                panic!("TtSequence in quote!")
```

Comment:
Could you move this to the top as

```rust
if !matcher {
    panic!("TtSequence in quote!");
}
```

and then un-indent the bulk of the code?